include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+    ObjectIndex(ObjectIndex const&) = delete;
+    ObjectIndex& operator=(ObjectIndex const&) = delete;
+
+    /// The non-trivial destructor is needed to close and release resources used by the index.
+    ~ObjectIndex();
+
+    /// Modes for opening the index.
+    enum Mode {
+        READ,
+        WRITE
+    };
+
+    /**
+     * The factory method.
+     * 
+     * @return  A reference to the underlining implementation of the index
+     */
+    static ObjectIndex& instance() {
+        static ObjectIndex oi;
+        return oi;
+    }
+
+    /// @return A value of 'true' if the index is open or created.
+    bool isOpen() const { return _isOpen; }
+
+    /// @return A mode the index was open/created in.
+    Mode mode() const { return _mode; }
+
+    /**
+     * Open (or create) index for writing into a local file.
+     * 
+     * @param fileName  A path to a local file to be open/created.
+     * @param editor  A CSV "editor" which is used for formatting output products.
+     * @param idFieldName  The name of an output field representing object identifiers.
+     * @param chunkIdFieldName  The name of an output field representing chunk identifiers.
+     * @param subChunkIdFieldName  The name of an output field representing sub-chunk identifiers.
+     * @throw std::invalid_argument  Invalid specificaton of the fields.
+     * @throw std::runtime_error  If the file opening/creation operation failed for another reason.
+     */
+    void create(std::string const& fileName,
+                csv::Editor const& editor,
+                std::string const& idFieldName,
+                std::string const& chunkIdFieldName,
+                std::string const& subChunkIdFieldName);
+
+    /**
+     * Open the index.
+     * 
+     * Note that parameter 'dialect' should be normally the same one which is
+     * used for making output products of the partitioner. This is based on an assumption
+     * that all products made by the partitioner in a context of a given catalog would be
+     * partitoned in the same way. If that's not the case 
+     *
+     * @param url  An index specification.
+     * @param dialect  A dialect for parsing index specifications.
+     * @throw std::invalid_argument  Invalid specificaton of the index.
+     * @throw std::runtime_error  Failed to open the index.
+     */
+    void open(std::string const& url,
+              csv::Dialect const& dialect);
+
+    /// Close the index and release all resources associated with it. 
+    void close();
+
+    /**
+     * Write a record into the index.
+     * 
+     * @param id  An identifier of an object.
+     * @param location  A location of the object (including chunk and sub-chunk identifiers).
+     * @throw std::logic_error  If the index was not created or if it's not open in Mode::WRITE.
+     * @throw std::invalid_argument  Invalid values of the input parameters.
+     */
+    void write(std::string const& id, ChunkLocation const& location);
```

Comment:
Yes! From the partitioner's point of view, anything but `ra`, `dec`, `chunkId` and `subChunkId` is a string (a value from the corresponding column/field of the input `CSV`/`TSV` file). As it should be clear from the implementation of the "secondary" index, the index is stored as a`CSV`/`TSV` file which is a subset of what's stored in the output chunk files.
The same would work for MySQL and Redis.
If a context of your question is multi-field complex object identifier then a serialized `CSV` (into a string) would also work here. So, the string-based approach seems to be "future-proof". Of course, such complex key would require making changes in many areas, of which the partitioner would be the smallest problem (just think about the "secondary" index within Qserv).

Reply:
But in MySQL secondary table the objectId type should be the same as in the actual Object schema. I know that we use CSV today to load data into MySQL secondary index but if we ever change that it would be safer to use actual type for director key.

Reply:
We will NEVER do that (loading any data into Qserv's MySQL from the partitioner). There are many reasons for that:
* we don't want to have any date implicitly ingested into internal data structures of Qserv before starting any super-transactions
* loading into MySQL w/o knowing a specific identifier of the super-transaction would't let the Ingests system to associate input products (chunk contributions) and the :secondary" index contributions.
* such direct dependency onto an internal MySQL server of Qserv may not be even possible (due to all sorts of firewall and security limitations)

The MySQL index implementation option mentioned earlier would be based on a dedicated MySQL service which would be setup to support the partitioning. Same is for Redis. In both cases assuming that the "objectId" is a string would work well. And tis would also make the operation agnostic to the actual type of the index. In the end, unlike `ra` and `dec`, we're not really interested in an actual value of the `objectId`. All we need to know is to establish a reliable association between a row from the director table with the corresponding rows in the dependent/child tables.

Moreover, I'm very pessimistic about using either of the above mentioned server technologies for partitioning. The high latency and low bandwidth of those would kill the performance of the partitioner.

Reply:
I think I disagree with all of that. Representing all possible data as stings may work (in some limited sense of "work") but I can imagine that knowing actual type of data can help in optimizing things. Indexing integers is probably much more efficient operation that indexing arbitrary strings.

Reply:
Okay, let me give you an example of what might happen if we used a MySQL-based "secondary" index for partitioning child tables. Let's assume we have have a MySQL service which is capable of processing 1000 `SELECT` queries per second. Let's also assume that each such query would return `chunkId` & `subChunkId` for a given `objectId` (forget for a moment about its type). Let's now assume that we're partitioning some "narrow" table like LSST's `ForcedSource` which has 100 columns of 4 bytes each. Let's also assume we have 10 rows in this table which have the sae value of the `objectId` field. Therefore, in order to process 10 rows:
```
10 * 100 * 4 = 4 KBytes
```
we would have to wait for 1 millisecond. This will result in the miserable performance of the partitioner:
```
1000 indexes/sec * 4 KB = 4 MB/sec
```
You can do the rest of this trivial math for 1 PB dataset of LSST DR2: `250,000,000 seconds` (25 years) just to partition.
Now about Redis. In my experiments with Redis I observed a performance in the range of 100,000 to 1,000,000 queries per second. This would speedup the partitioner to end up its job within a range of 3 months to  9 days. (`400 MB/s` or `4 GB/s`). That would be horrible.
The only architecture which would work for pre-processing the LSST-scale data would be *streaming*, not the single-point-of-service one.
I should also mention that the file-based index approach has its own problems and limitations. Consider a scenario where one would have to process 1 billion objects. In this case the total size of the index would be roughly `28 GB` (based on my experiments with the *Wise* catalogs and the string-based objectId). This means the partitioner's process would have to read into its memory `28 GB` of the index data  even before doing anything useful. This would also kill the performance of the application.
Hence, realistically speaking the file-based indexing approach would only work for scenarios where relatively small amounts of the director and dependent tables' data are reasonably limited while ensuing the referential integrity between these products.
In my opinion, this problem doesn't have a universal solution. All we have would be engineering compromises to be found in specific application contexts.